### PR TITLE
Support For Saving Location Entity mapping After Sage Intacct Connection

### PR DESCRIPTION
**Pull request by NileshPant1999**

### CODE REVIEW

File: apps/mappings/models.py [hunk 1]
```diff
@@ -6,6 +6,21 @@
...
 
 
+class LocationEntityMapping(models.Model):
+    """
+    Location Entity Mapping
+    """
+    id = models.AutoField(primary_key=True)
+    location_entity_name = models.CharField(max_length=255, help_text='SageIntacct Location Entity Name')
+    country_name = models.CharField(max_length=255, help_text='SageIntacct Location Entities Country', null=True)
+    destination_id = models.CharField(max_length=255, help_text='SageIntacct location entity id')
```
Comment (ashwin1111) [inline apps/mappings/models.py:16]: shouldn't this be `null=True`?
How are we handling top level ids?

File: apps/mappings/serializers.py [hunk 2]
```diff
@@ -10,3 +10,11 @@ class GeneralMappingSerializer(serializers.ModelSerializer):
...
         model = GeneralMapping
         fields = '__all__'
+
+class LocationEntityMappingSerializer(serializers.ModelSerializer):
+    """
+    Location Entity Mappings group serializer
+    """
+    class Meta:
+        model = LocationEntityMapping
+        fields = '__all__'
```
Comment (ashwin1111) [inline apps/mappings/serializers.py:20]: new line

File: apps/sage_intacct/helpers.py [hunk 3]
```diff
@@ -56,7 +56,7 @@ def sync_dimensions(si_credentials: SageIntacctCredential, workspace_id: int, di
         dimensions = [
             'locations', 'customers', 'departments', 'projects', 'expense_payment_types',
             'classes', 'charge_card_accounts','payment_accounts', 'vendors', 'employees', 'accounts',
-            'expense_types', 'items', 'user_defined_dimensions'
+            'expense_types', 'items', 'user_defined_dimensions', 'location_entities'
```
Comment (ashwin1111) [inline apps/sage_intacct/helpers.py:59]: Why do we need to sync `location_entities` on every sync? We can just sync them once during post of sage intacct creds. Would be nice if we can use a post save signal.

File: apps/sage_intacct/utils.py [hunk 4]
```diff
@@ -263,7 +263,11 @@ def sync_location_entities(self):
                 'attribute_type': 'LOCATION_ENTITY',
                 'display_name': 'location entity',
                 'value': location_entity['NAME'],
-                'destination_id': location_entity['LOCATIONID']
+                'destination_id': location_entity['LOCATIONID'],
+                'detail': {
+                    'country': location_entity['OPCOUNTRY'],
+                    'currency': location_entity['CURRENCY']
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:269]: Just a question, where will this be impacted?

File: apps/workspaces/views.py [hunk 5]
```diff
@@ -65,7 +65,7 @@ def post(self, request):
 
             FyleCredential.objects.update_or_create(
                 refresh_token=auth_tokens.refresh_token,
-                workspace_id=workspace.id
+                workspace_id=workspace.id,
```
Comment (ashwin1111) [inline apps/workspaces/views.py:68]: remove `,`

File: apps/workspaces/views.py [hunk 6]
```diff
@@ -166,6 +166,10 @@ def post(self, request, **kwargs):
             workspace.fyle_org_id = org_id
             workspace.save()
 
+            fyle_connector = FyleConnector(refresh_token=refresh_token, workspace_id=kwargs['workspace_id'])
+
+            cluster_domain = fyle_connector.get_cluster_domain()['cluster_domain']
```
Comment (ashwin1111) [inline apps/workspaces/views.py:171]: I don't see `cluster_domain` being used

**Comment by ashwin1111** [comment]
@NileshPant1999 let's also try to add tests along with new features

File: apps/workspaces/views.py (hunk 6 above)
Comment (NileshPant1999) [reply apps/workspaces/views.py:171]: will remove this, came from tax changes

File: apps/sage_intacct/utils.py (hunk 4 above)
Comment (NileshPant1999) [reply apps/sage_intacct/utils.py:269]: was trying to use both to check weather to disable the import tax or not will remove currency now as we are not using location entity with country set as null

File: apps/sage_intacct/helpers.py (hunk 3 above)
Comment (NileshPant1999) [reply apps/sage_intacct/helpers.py:59]: will remove it from here

File: apps/mappings/models.py (hunk 1 above)
Comment (NileshPant1999) [reply apps/mappings/models.py:16]: I thought of adding a dummy a value like a dummy string but null will also work will change it to null=True

File: apps/mappings/signals.py [hunk 7]
```diff
@@ -15,10 +15,26 @@
...
+@receiver(post_save, sender=LocationEntityMapping)
+def run_post_location_entity_mappings_triggers(sender, instance: LocationEntityMapping, **kwargs):
+    """
+    :param sender: Sender Class
+    :param instance: Row instance of Sender Class
+    :return: None
+    """
+
+    sage_intacct_credentials = SageIntacctCredential.objects.get(workspace_id=int(instance.workspace_id))
+    sync_dimensions(sage_intacct_credentials, int(instance.workspace_id))
```
Comment (ashwin1111) [inline apps/mappings/signals.py:36]: let's do this in UI, this will block client

File: apps/sage_intacct/utils.py [hunk 8]
```diff
@@ -39,15 +39,20 @@ def __init__(self, credentials_object: SageIntacctCredential, workspace_id: int)
         decrypted_password = cipher_suite.decrypt(credentials_object.si_user_password.encode('utf-8')).decode('utf-8')
 
         # TODO: Cache general_mappings
-        general_mappings = GeneralMapping.objects.filter(workspace_id=workspace_id).first()
+        location_entity_mapping = LocationEntityMapping.objects.filter(workspace_id=workspace_id).first()
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:42]: add destination id shouldn't be 'top_level' in orm filter. And we can solve this in a single line
```
        self.connection = SageIntacctSDK(
            sender_id=sender_id,
            sender_password=sender_password,
            user_id=credentials_object.si_user_id,
            company_id=credentials_object.si_company_id,
            user_password=decrypted_password,
            entity_id=location_entity_mapping.destination_id if location_entity_mapping else None
        )

File: apps/workspaces/signals.py [hunk 9]
```diff
@@ -4,11 +4,12 @@
 from django.db.models.signals import post_save
 from django.dispatch import receiver
 
+from fyle_accounting_mappings.models import DestinationAttribute
```
Comment (ashwin1111) [inline apps/workspaces/signals.py:7]: not needed

File: apps/workspaces/signals.py [hunk 10]
```diff
@@ -4,11 +4,12 @@
...
 from apps.fyle.helpers import add_expense_id_to_expense_group_settings
 from apps.sage_intacct.helpers import schedule_payment_sync
 from apps.mappings.helpers import schedule_or_delete_auto_mapping_tasks
 
-from .models import Configuration
+from .models import Configuration, SageIntacctCredential
```
Comment (ashwin1111) [inline apps/workspaces/signals.py:12]: not needed

**Review by ashwin1111 — APPROVED** [review]

**Pull request merged**